chainstore.go

```diff
@@ -1,147 +1,231 @@
 func New(stores ...Store) Store {
-	return &Chain{stores, false}
-	// TODO: make the chain..
-	// call Open(), but in case of error..?
+	c := &Chain{
+		stores: make([]*storeWrapper, 0, len(stores)),
+	}
+	for _, s := range stores {
+		c.stores = append(c.stores, &storeWrapper{Store: s})
+	}
+	return c
 }
 
-func Async(stores ...Store) Store {
-	return &Chain{stores, true}
-}
+// Open opens all the stores.
+func (c *Chain) Open() error {
 
-func (c *Chain) Open() (err error) {
-	for _, s := range c.stores {
-		err = s.Open()
-		if err != nil {
-			return // return first error that comes up
-		}
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	return
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, c *Chain, i int) {
+			s := c.stores[i]
+
+			s.setErr(s.Open())
+
+			wg.Done()
+		}(&wg, c, i)
+	}
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Close() (err error) {
-	for _, s := range c.stores {
-		err = s.Close()
-		// TODO: we shouldn't stop on first error.. should keep trying to close
-		// and record errors separately
-		if err != nil {
-			return
-		}
+// Close closes all the stores.
+func (c *Chain) Close() error {
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Close failed due to a previous error: %q", err)
+	}
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, s *storeWrapper) {
+			s.setErr(s.Close())
+			wg.Done()
+		}(&wg, c.stores[i])
 	}
-	return
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Put(key string, val []byte) (err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Put(ctx context.Context, key string, val []byte) (err error) {
+	if !isValidKey(key) {
 		return ErrInvalidKey
 	}
 
-	fn := func() (err error) {
-		for _, s := range c.stores {
-			err = s.Put(key, val)
-			if err != nil {
-				return
-			}
-		}
-		return
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	if c.async {
-		go fn()
-	} else {
-		err = fn()
+
+	fn := func(s Store) error {
+		return s.Put(ctx, key, val)
 	}
-	return
+
+	return c.doWithContext(ctx, fn)
 }
 
-func (c *Chain) Get(key string) (val []byte, err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Get(ctx context.Context, key string) (val []byte, err error) {
+	if !isValidKey(key) {
 		return nil, ErrInvalidKey
 	}
 
-	for i, s := range c.stores {
-		val, err = s.Get(key)
-		if err != nil {
-			return
-		}
+	if err := c.firstErr(); err != nil {
+		return nil, fmt.Errorf("Open failed due to a previous error: %q", err)
+	}
 
-		if len(val) > 0 {
-			if i > 0 {
-				// put the value in all other stores up the chain
-				fn := func() {
-					for n := i - 1; n >= 0; n-- {
-						c.stores[n].Put(key, val) // errors..?
-					}
-				}
-				// if c.async { } else { } ....?
-				go fn()
+	errCh := make(chan error, 1)
+
+	go func() {
```

Comment:
I don't think this goroutine and errCh helps with anything. It will be fetching from all the stores sequentially anyways - so why in background? And how do you close this goroutine on `ctx.Done()`?

How about
```go

nextStore := make(chan Store, 0, len(c.stores))
for _, store := range c.stores {
      nextStore <- store
}
close(nextStore) // this will trigger !ok condition after all stores are received

putBack := make(chan Store, 0, len(c.stores))

for {
    select {
        case <-ctx.Done():
            return nil, ctx.Err()
        case store, ok := <-nextStore:
            if !ok {
                return nil, ErrNotFound
            }
            val, err := store.Get(ctx, key)
            if err != nil { // not found in this store
                putBack <- store
                continue // next store
            }
            for _, store := range putBack {
                go store.Put(ctx, key, val)
            }
            return val, nil
    }
}
```